### PR TITLE
Lms 2406 notification email content

### PR DESCRIPTION
**Pull request by stephen-oh**
**Description**
- Changing the email notifications to list only counts of each error type. Using mustache templates for email content. Logging only 1 row per user per notification email and changes to the notification threshold.

**Related Issues**
- LMS-2406

**PR Type**
- Enhancement

### CODE REVIEW

File: classes/output/renderer.php [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace logstore_xapi\output;
+
+defined('MOODLE_INTERNAL') || die();
+
+use plugin_renderer_base;
+
+class renderer extends plugin_renderer_base {
```
Comment (lzabo) [inline classes/output/renderer.php:32]: the name should be logstore_xapi_renderer

Comment (stephen-oh) [reply classes/output/renderer.php:32]: I'm using namespace, see line 26

Comment (lzabo) [reply classes/output/renderer.php:32]: https://docs.moodle.org/dev/Overriding_a_renderer

> There is also a routine way to name renderers much like the process for naming a theme's overriding renderer.
> 
> It should always be made up of the following three components joined by underscores:
> 
> component This is the component the renderer belongs too, e.g. block, mod, or core.
> name This is the name of what ever the component is, e.g. forum, workshop, or wiki.
> renderer They should always be completed by the word renderer.

Comment (stephen-oh) [reply classes/output/renderer.php:32]: https://docs.moodle.org/dev/Overriding_a_renderer#Namespaces

> Renderers can also be used with code that uses namespaces. When creating or overriding new renderers with namespaces, be sure to use "output" as the second level namespace to comply with the rules for namespaces.

Comment (lzabo) [reply classes/output/renderer.php:32]: Yes, thats true. 

Renderers can also be used with code that uses namespaces.

Comment (lzabo) [reply classes/output/renderer.php:32]: So you need to use namespace and follow the naming convention.

**Review by gordonmacqueen-lp — APPROVED** [review]

**Review by mlynn-lp — APPROVED** [review]

File: classes/task/sendfailednotifications_task.php [hunk 2]
```diff
@@ -47,93 +45,70 @@ public function get_name() {
     }
 
     /**
-     * Get failed rows.
+     * Get the last id of the new notifications to send, if there are no notifications to send then return false.
      *
-     * @return array
+     * @return array|bool
```
Comment (gordonmacqueen-lp) [inline classes/task/sendfailednotifications_task.php]: I think this should now be `int|bool`

File: classes/task/sendfailednotifications_task.php [hunk 3]
```diff
@@ -47,93 +45,70 @@ public function get_name() {
...
     }
 
     /**
      * Send email using email_to_user.
      *
-     * @param array $failedrows an array of failed rows from logstore_xapi_failed_log
      * @param string $message email message
      * @param string $subject email subject
      * @param object $user user to receive email
+     * @param $lastfailedid
```
Comment (gordonmacqueen-lp) [inline classes/task/sendfailednotifications_task.php]: int?

**Review by gordonmacqueen-lp — CHANGES_REQUESTED** [review]

File: classes/task/sendfailednotifications_task.php [hunk 4]
```diff
@@ -47,93 +45,70 @@ public function get_name() {
...
+        $lastnotification = $DB->get_record_sql('SELECT MAX(failedlogid) AS id FROM {logstore_xapi_notif_sent_log}');
+
+        if ($lastnotification) {
+            $lastnotificationid = $lastnotification->id;
+        }
+
+        $sql = 'SELECT MAX(id) AS id
+                  FROM {logstore_xapi_failed_log}
+                 WHERE id > :lastsentid
+                HAVING COUNT(id) > :threshold';
```
Comment (gordonmacqueen-lp) [inline classes/task/sendfailednotifications_task.php]: `>=`

**Review by gordonmacqueen-lp — APPROVED** [review]

**Pull request merged**